File: plugins/src/android/OptimoveSDKPlugin.java
package com.optimove.android.cordova;

import android.content.Context;
import androidx.annotation.Nullable;

import java.net.URL;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.List;
import java.util.Locale;
import java.util.TimeZone;

import com.optimove.android.Optimove;
import com.optimove.android.optimobile.InAppDeepLinkHandlerInterface;
import com.optimove.android.optimobile.InAppInboxItem;
import com.optimove.android.optimobile.InAppInboxSummary;
import com.optimove.android.optimobile.OptimoveInApp;
import com.optimove.android.optimobile.PushMessage;

import org.apache.cordova.CordovaInterface;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CallbackContext;
import org.apache.cordova.PluginResult;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

public class OptimoveSDKPlugin extends CordovaPlugin {

    private static final String SET_HANDLERS_CALLBACK_CONTEXT = "setHandlersCallBackContext";
    private static final String SET_USER_ID = "setUserId";
    private static final String SET_USER_EMAIL = "setUserEmail";
    private static final String REPORT_EVENT = "reportEvent";
    private static final String REPORT_SCREEN_VISIT = "reportScreenVisit";
    private static final String REGISTER_USER = "registerUser";
    private static final String GET_VISITOR_ID = "getVisitorId";
    private static final String GET_CURRENT_USER_IDENTIFIER = "getCurrentUserIdentifier";
    private static final String PUSH_REGISTER = "pushRegister";
    private static final String IN_APP_UPDATE_CONSENT = "inAppUpdateConsent";
    private static final String IN_APP_GET_INBOX_ITEMS = "inAppGetInboxItems";
    private static final String IN_APP_MARK_ALL_INBOX_ITEMS_AS_READ = "inAppMarkAllInboxItemsAsRead";
    private static final String IN_APP_MARK_AS_READ = "inAppMarkAsRead";
    private static final String IN_APP_GET_INBOX_SUMMARY = "inAppGetInboxSummary";
    private static final String IN_APP_PRESENT_INBOX_MESSAGE = "inAppPresentInboxMessage";
    private static final String IN_APP_DELETE_INBOX_MESSAGE = "inAppDeleteMessageFromInbox";
    private static final String CHECK_IF_PENDING_PUSH_EXISTS = "checkIfPendingPushExists";
    @Nullable
    static CallbackContext jsCallbackContext;
    @Nullable
    static PushMessage pendingPush;
    @Nullable
    static String pendingActionId;
    static CordovaInterface sCordova;

    @Override
    public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
        switch (action) {

        case SET_HANDLERS_CALLBACK_CONTEXT:
            setHandlersCallBackContext(callbackContext);
            return true;

        case SET_USER_ID:
            this.setUserId(args, callbackContext);
            return true;

        case SET_USER_EMAIL:
            this.setUserEmail(args, callbackContext);
            return true;

        case REPORT_EVENT:
            this.reportEvent(args, callbackContext);
            return true;

        case REPORT_SCREEN_VISIT:
            reportScreenVisit(args, callbackContext);
            return true;

        case REGISTER_USER:
            this.registerUser(args, callbackContext);
            return true;

        case GET_VISITOR_ID:
            this.getVisitorId(callbackContext);
            return true;

        case GET_CURRENT_USER_IDENTIFIER:
            this.getCurrentUserIdentifier(callbackContext);
            return true;

        case PUSH_REGISTER:
            this.pushRegister(callbackContext);
            return true;

        case IN_APP_UPDATE_CONSENT:
            this.inAppUpdateConsent(args, callbackContext);
            return true;

        case IN_APP_GET_INBOX_ITEMS:
            cordova.getThreadPool().execute(() -> OptimoveSDKPlugin.this.inAppGetInboxItems(callbackContext));
            return true;

        case IN_APP_MARK_ALL_INBOX_ITEMS_AS_READ:
            cordova.getThreadPool().execute(() -> OptimoveSDKPlugin.this.inAppMarkAllInboxItemsAsRead(callbackContext));
            return true;

        case IN_APP_MARK_AS_READ:
            cordova.getThreadPool().execute(() -> OptimoveSDKPlugin.this.inAppMarkAsRead(args, callbackContext));
            return true;

        case IN_APP_GET_INBOX_SUMMARY:
            this.inAppGetInboxSummary(callbackContext);
            return true;

        case IN_APP_PRESENT_INBOX_MESSAGE:
            cordova.getThreadPool()
                    .execute(() -> OptimoveSDKPlugin.this.inAppPresentInboxMessage(args, callbackContext));
            return true;

        case IN_APP_DELETE_INBOX_MESSAGE:
            cordova.getThreadPool()
                    .execute(() -> OptimoveSDKPlugin.this.inAppDeleteMessageFromInbox(args, callbackContext));
            return true;
        case CHECK_IF_PENDING_PUSH_EXISTS:
            this.checkIfPendingPushExists();
            return true;
        }

        return false;
    }

    private void inAppDeleteMessageFromInbox(JSONArray args, CallbackContext callbackContext) {
        int messageId = args.optInt(0, -1);
        InAppInboxItem item = getInboxItemById(messageId);
        if (item == null) {
            callbackContext.error("Message not found or not available");
        }
        boolean result = OptimoveInApp.getInstance().deleteMessageFromInbox(item);
        if (result) {
            callbackContext.success();
        } else {
            callbackContext.error("Failed to delete inbox item");
        }
    }

    private void inAppPresentInboxMessage(JSONArray args, CallbackContext callbackContext) {
        int messageId = args.optInt(0, -1);
        InAppInboxItem item = getInboxItemById(messageId);
        if (item == null) {
            callbackContext.error("Message not found or not available");
            return;
        }
        OptimoveInApp.InboxMessagePresentationResult result = OptimoveInApp.getInstance().presentInboxMessage(item);
        if (result == OptimoveInApp.InboxMessagePresentationResult.PRESENTED) {
            callbackContext.success();
        } else {
            callbackContext.error("Failed to present message");
        }
    }

    private void inAppGetInboxSummary(CallbackContext callbackContext) {
        OptimoveInApp.getInstance().getInboxSummaryAsync((InAppInboxSummary summary) -> {
            if (summary == null) {
                callbackContext.error("Could not get inbox summary");
                return;
            }

            try {
                JSONObject res = new JSONObject();
                res.put("totalCount", summary.getTotalCount());
                res.put("unreadCount", summary.getUnreadCount());
                callbackContext.success(res);
            } catch (JSONException e) {
                e.printStackTrace();
                callbackContext.error(e.getMessage());
            }
        });
    }

    private void inAppMarkAsRead(JSONArray args, CallbackContext callbackContext) {
        int messageId = args.optInt(0, -1);
        InAppInboxItem item = getInboxItemById(messageId);
        if (item == null) {
            callbackContext.error("Message not found or not available");
        }
        boolean result = OptimoveInApp.getInstance().markAsRead(item);
        if (result) {
            callbackContext.success();
        } else {
            callbackContext.error("Failed to mark message as read");
        }
    }

    private void setUserId(JSONArray args, CallbackContext callbackContext) {
        try {
            String userId = args.getString(0);
            Optimove.getInstance().setUserId(userId);
            callbackContext.success();
        } catch (Exception e) {
            callbackContext.error(e.getMessage());
            e.printStackTrace();
        }
    }

    private void setUserEmail(JSONArray args, CallbackContext callbackContext) {
        try {
            String email = args.getString(0);
            Optimove.getInstance().setUserEmail(email);
            callbackContext.success();
        } catch (Exception e) {
            callbackContext.error(e.getMessage());
            e.printStackTrace();
        }
    }

    private void reportEvent(JSONArray args, CallbackContext callbackContext) {
        try {
            String eventName = args.getString(0);
            JSONObject params = args.optJSONObject(1);
            if (params == null) {
                Optimove.getInstance().reportEvent(eventName);
            } else {
                Optimove.getInstance().reportEvent(eventName, JsonUtils.toMap(params));
            }
            callbackContext.success();
        } catch (Exception e) {
            callbackContext.error(e.getMessage());
            e.printStackTrace();
        }
    }

    private void reportScreenVisit(JSONArray args, CallbackContext callbackContext) {
        try {
            String screenName = args.getString(0);

            if (args.isNull(1)) {
                Optimove.getInstance().reportScreenVisit(screenName);
            } else {
                String screenCategory = args.optString(1);
                Optimove.getInstance().reportScreenVisit(screenName, screenCategory);
            }
            callbackContext.success();
        } catch (Exception e) {
            callbackContext.error(e.getMessage());
            e.printStackTrace();
        }
    }

    private void registerUser(JSONArray args, CallbackContext callbackContext) {
        try {
            String userId = args.getString(0);
            String userEmail = args.getString(1);
            Optimove.getInstance().registerUser(userId, userEmail);
            callbackContext.success();
        } catch (Exception e) {
            callbackContext.error(e.getMessage());
            e.printStackTrace();
        }
    }

    private void getVisitorId(CallbackContext callbackContext) {
        String visitorId = Optimove.getInstance().getVisitorId();
        if (visitorId != null) {
            callbackContext.success(visitorId);
        } else {
            callbackContext.error("visitor id is null");
        }
    }

    private void getCurrentUserIdentifier(CallbackContext callbackContext) {
        String currentUserIdentifier = Optimove.getInstance().getCurrentUserIdentifier();
        if (currentUserIdentifier != null) {
            callbackContext.success(currentUserIdentifier);
        } else {
            callbackContext.error("current user identifier is null");
        }
    }

    private void setHandlersCallBackContext(CallbackContext callbackContext) {
        jsCallbackContext = callbackContext;
        PluginResult result = new PluginResult(PluginResult.Status.OK);
        result.setKeepCallback(true);
        callbackContext.sendPluginResult(result);
    }

    private void checkIfPendingPushExists() {
        if (null != pendingPush) {
            OptimoveSDKPlugin.sendMessageToJs("pushOpened",
                    PushReceiver.pushMessageToJsonObject(pendingPush, pendingActionId));
            pendingPush = null;
            pendingActionId = null;
        }
    }

    static boolean sendMessageToJs(String type, JSONObject data) {
        if (null == jsCallbackContext) {
            return false;
        }

        JSONObject message = new JSONObject();
        try {
            message.put("type", type);
            message.put("data", data);
        } catch (JSONException e) {
            e.printStackTrace();
            return false;
        }

        PluginResult result = new PluginResult(PluginResult.Status.OK, message);
        result.setKeepCallback(true);
        jsCallbackContext.sendPluginResult(result);

        return true;
    }

    private void pushRegister(CallbackContext callbackContext) {
        try {
            Optimove.getInstance().pushRegister();
            callbackContext.success();
        } catch (Exception e) {
            callbackContext.error(e.getMessage());
            e.printStackTrace();
        }

    }

    private void inAppUpdateConsent(JSONArray args, CallbackContext callbackContext) {
        final boolean consented;

        try {
            consented = args.getBoolean(0);
            OptimoveInApp.getInstance().updateConsentForUser(consented);
            callbackContext.success();
        } catch (Exception e) {
            e.printStackTrace();
            callbackContext.error(e.getMessage());
        }
    }

    private void inAppGetInboxItems(CallbackContext callbackContext) {
        SimpleDateFormat formatter = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.US);
        formatter.setTimeZone(TimeZone.getTimeZone("UTC"));

        List<InAppInboxItem> items = OptimoveInApp.getInstance().getInboxItems();
        JSONArray results = new JSONArray();
        try {
            for (InAppInboxItem item : items) {
                JSONObject mapped = new JSONObject();

                mapped.put("id", item.getId());
                mapped.put("title", item.getTitle());
                mapped.put("subtitle", item.getSubtitle());
                mapped.put("isRead", item.isRead());
                mapped.put("sentAt", formatter.format(item.getSentAt()));

                Date availableFrom = item.getAvailableFrom();
                Date availableTo = item.getAvailableTo();
                Date dismissedAt = item.getDismissedAt();
                mapped.put("data", item.getData());

                URL imageUrl = item.getImageUrl();
                mapped.put("imageUrl", imageUrl == null ? null : imageUrl.toString());

                if (null == availableFrom) {
                    mapped.put("availableFrom", "");
                } else {
                    mapped.put("availableFrom", formatter.format(availableFrom));
                }

                if (null == availableTo) {
                    mapped.put("availableTo", "");
                } else {
                    mapped.put("availableTo", formatter.format(availableTo));
                }

                if (null == dismissedAt) {
                    mapped.put("dismissedAt", "");
                } else {
                    mapped.put("dismissedAt", formatter.format(dismissedAt));
                }

                results.put(mapped);
            }
        } catch (JSONException e) {
            e.printStackTrace();
            callbackContext.error(e.getMessage());
            return;
        }

        callbackContext.success(results);
    }

    private void inAppMarkAllInboxItemsAsRead(CallbackContext callbackContext) {
        boolean result = OptimoveInApp.getInstance().markAllInboxItemsAsRead();
        if (result) {
            callbackContext.success();
        } else {
            callbackContext.error("Failed to mark all messages as read");
        }
    }

    private InAppInboxItem getInboxItemById(int id) {
        if (id == -1) {
            return null;
        }
        List<InAppInboxItem> inboxItems = OptimoveInApp.getInstance().getInboxItems();
        for (InAppInboxItem item : inboxItems) {
            if (item.getId() == id) {
                return item;
            }
        }
        return null;
    }

    static class InAppDeepLinkHandler implements InAppDeepLinkHandlerInterface {

        @Override
        public void handle(Context context, InAppButtonPress buttonPress) {
            sendMessageToJs("inAppDeepLinkPressed", buttonPress.getDeepLinkData());
        }
    }

    static class InboxUpdatedHandler implements OptimoveInApp.InAppInboxUpdatedHandler {
        @Override
        public void run() {
            sendMessageToJs("inAppInboxUpdated", null);
        }
    }
}

File: cordova-sdk/src/ios/OptimoveSDKPlugin.swift
import OptimoveSDK
import NotificationCenter

enum InAppConsentStrategy: String {
    case autoEnroll = "auto-enroll"
    case explicitByUser = "explicit-by-user"
    case disabled = "in-app-disabled"
}

@objc(Optimove_Cordova) class OptimoveSDKPlugin : CDVPlugin {
    private static var optimovePluginInstance: OptimoveSDKPlugin!
    private static var cordovaCommand: CDVInvokedUrlCommand? = nil

    private static let optimoveCredentialsKey = "optimoveCredentials"
    private static let optimoveMobileCredentialsKey = "optimoveMobileCredentials"
    private static let inAppConsentStrategy = "optimoveInAppConsentStrategy"
    private static let enableDeferredDeepLinking = "optimoveEnableDeferredDeepLinking"
    private static let cname = "optimoveDdlCname"

    private static var pendingPush: PushNotification? = nil
    private static var pendingDdl: DeepLinkResolution? = nil

    private static let sdkVersion = "1.0.1"
    private static let sdkTypeOptimoveCordova = 106
    private static let runtimeTypeCordova = 3

    // ========================== INITIALIZATION ==========================

    override func pluginInitialize() {
        OptimoveSDKPlugin.optimovePluginInstance = self
    }

    @objc(didFinishLaunching:cdvVersion:)
    static func didFinishLaunching(notification: Notification, cdvVersion: String) {

        let configPath = Bundle.main.path(forResource: "optimove", ofType: "plist")

        guard let configPath = configPath else {
            print("optimove.plist NOT FOUND")
            return
        }

        guard let configValues: [String: String] = NSDictionary(contentsOfFile: configPath) as? [String: String] else {
            print("optimove.plist IS NOT VALID")
            return
        }

        guard let builder = getConfigBuilder(configValues: configValues) else{
            return
        };

        builder.setPushOpenedHandler(pushOpenedHandlerBlock: { notification in
            guard let pluginInstance = OptimoveSDKPlugin.optimovePluginInstance, let _ = OptimoveSDKPlugin.cordovaCommand else {
                pendingPush = notification

                return
            }

            pluginInstance.sendMessageToJs(type: "pushOpened", data: getPushNotificationMap(pushNotification: notification))
        })

        if #available(iOS 10, *) {
            builder.setPushReceivedInForegroundHandler(pushReceivedInForegroundHandlerBlock: { notification, completionHanlder in
                OptimoveSDKPlugin.optimovePluginInstance?.sendMessageToJs(type: "pushReceived", data: getPushNotificationMap(pushNotification: notification))

                completionHanlder(UNNotificationPresentationOptions.alert)
            })
        }

        switch(configValues[inAppConsentStrategy]){
            case InAppConsentStrategy.autoEnroll.rawValue:
                builder.enableInAppMessaging(inAppConsentStrategy:OptimoveSDK.InAppConsentStrategy.autoEnroll);
                break
            case InAppConsentStrategy.explicitByUser.rawValue:
                builder.enableInAppMessaging(inAppConsentStrategy:OptimoveSDK.InAppConsentStrategy.explicitByUser);
                break
            case InAppConsentStrategy.disabled.rawValue:
                break
            default:
                print("Invalid inApp consent strategy")
                return
        }

        builder.setInAppDeepLinkHandler(inAppDeepLinkHandlerBlock: { data in
            OptimoveSDKPlugin.optimovePluginInstance?.sendMessageToJs(type: "inAppDeepLink", data: getInappButtonPressMap(inAppButtonPress: data))
        })

        if (configValues[enableDeferredDeepLinking] != nil) {
            let ddlHandler: DeepLinkHandler = { deepLinkResolution in
                guard let pluginInstance = OptimoveSDKPlugin.optimovePluginInstance, let _ = OptimoveSDKPlugin.cordovaCommand else {
                    pendingDdl = deepLinkResolution

                    return
                }

                pluginInstance.sendMessageToJs(type: "deepLink", data: getDdlResolutionMap(deepLinkResolution: deepLinkResolution))
            }

            if (configValues[cname] != nil){
                builder.enableDeepLinking(cname: configValues[cname], ddlHandler)
            }
            else{
                builder.enableDeepLinking(ddlHandler)
            }
        }

        overrideInstallInfo(builder: builder, cdvVersion:cdvVersion)

        let config = builder.build()

        Optimove.initialize(with: config)

        OptimoveInApp.setOnInboxUpdated(inboxUpdatedHandlerBlock: {
            OptimoveSDKPlugin.optimovePluginInstance?.sendMessageToJs(type: "inAppInboxUpdated", data: nil)
        })
    }

    static func overrideInstallInfo(builder: OptimoveConfigBuilder, cdvVersion: String) -> Void {
        let runtimeInfo: [String : AnyObject] = [
            "id": runtimeTypeCordova as AnyObject,
            "version": cdvVersion as AnyObject,
        ]

        let sdkInfo: [String : AnyObject] = [
            "id": sdkTypeOptimoveCordova as AnyObject,
            "version": sdkVersion as AnyObject,
        ]

        builder.setRuntimeInfo(runtimeInfo: runtimeInfo);
        builder.setSdkInfo(sdkInfo: sdkInfo);

        var isRelease = true
        #if DEBUG
            isRelease = false
        #endif
        builder.setTargetType(isRelease: isRelease);
    }

    static func getConfigBuilder(configValues: [String: String]) -> OptimoveConfigBuilder? {
        var optimoveCredentials: String? = nil
        if let val = configValues[optimoveCredentialsKey] {
            if (!val.isEmpty){
                optimoveCredentials = val;
            }
        }

        var optimobileCredentials: String? = nil
        if let val = configValues[optimoveMobileCredentialsKey] {
            if (!val.isEmpty){
                optimobileCredentials = val;
            }
        }

        let builder = OptimoveConfigBuilder(optimoveCredentials: optimoveCredentials, optimobileCredentials: optimobileCredentials)

        return builder
    }

    // ========================== ASSOCIATION AND EVENTS ==========================

    @objc(reportEvent:)
    func reportEvent(command: CDVInvokedUrlCommand) {
        guard let name = command.arguments[0] as? String else {
            let pluginResult = CDVPluginResult(status: .error, messageAs: "name is invalid")
            self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
            return
        }

        var parameters: [String: Any] = [:]

        if let param = command.arguments[1] as? [String: Any] {
            parameters = param
        }

        Optimove.shared.reportEvent(name: name, parameters: parameters)
        let pluginResult = CDVPluginResult(status: .ok)
        self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
    }

    @objc(reportScreenVisit:)
    func reportScreenVisit(command: CDVInvokedUrlCommand) {
        guard let screenTitle = command.arguments.first as? String else {
            let pluginResult = CDVPluginResult(status: .error, messageAs: "screenTitle is invalid")
            self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
            return
        }

        let screenCategory = command.arguments[1] as? String

        Optimove.shared.reportScreenVisit(screenTitle: screenTitle, screenCategory: screenCategory)

        let pluginResult = CDVPluginResult(status: .ok)
        self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
    }

    @objc(registerUser:)
    func registerUser(command: CDVInvokedUrlCommand) {
        guard let id = command.arguments.first as? String else {
            let pluginResult = CDVPluginResult(status: .error, messageAs: "id is invalid")
            self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
            return
        }

        guard let  email = command.arguments[1] as? String else {
            let pluginResult = CDVPluginResult(status: .error, messageAs: "email is invalid")
            self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
            return
        }

        Optimove.shared.registerUser(sdkId: id, email: email)

        let pluginResult = CDVPluginResult(status: .ok)
        self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
    }

    @objc(setUserId:)
    func setUserId(command: CDVInvokedUrlCommand) {
        guard let id = command.arguments.first as? String else {
            let pluginResult = CDVPluginResult(status: .error, messageAs: "id is invalid")
            self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
            return
        }

        Optimove.shared.setUserId(id)

        let pluginResult = CDVPluginResult(status: .ok)
        self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
    }

    @objc(getVisitorId:)
    func getVisitorId(command: CDVInvokedUrlCommand) {
        let pluginResult = CDVPluginResult(status: .ok, messageAs: Optimove.getVisitorID())
        self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
    }

    @objc(signOutUser:)
    func signOutUser(command: CDVInvokedUrlCommand) {
        Optimove.shared.signOutUser()

        self.commandDelegate.send(CDVPluginResult(status: .ok), callbackId: command.callbackId)
    }

    @objc(setUserEmail:)
    func setUserEmail(command: CDVInvokedUrlCommand) {
        guard let email = command.arguments.first as? String else {
            let pluginResult = CDVPluginResult(status: .error, messageAs: "email is invalid")
            self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
            return
        }

        Optimove.shared.setUserEmail(email: email)

        let pluginResult = CDVPluginResult(status: .ok)
        self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
    }

    // ========================== MESSAGING ==========================

    @objc(setHandlersCallBackContext:)
    func setHandlersCallBackContext(command: CDVInvokedUrlCommand){
        OptimoveSDKPlugin.cordovaCommand = command

        let checkForPendingPush: Bool = command.arguments[0] as! Bool
        let checkForPendingDdl: Bool = command.arguments[1] as! Bool

        var sentPushResult = false
        var sentDdlResult = false
        if (checkForPendingPush){
            sentPushResult = maybeSendAndClearPendingPush(command: command)
        }

        if (checkForPendingDdl){
            sentDdlResult = maybeSendAndClearPendingDdl(command: command)
        }

        if (!sentPushResult && !sentDdlResult){
            let pluginResult:CDVPluginResult = CDVPluginResult.init(status: .ok)
            pluginResult.setKeepCallbackAs(true)
            self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
        }
    }

    @objc(clearContext:)
    func clearContext(command: CDVInvokedUrlCommand){
        OptimoveSDKPlugin.cordovaCommand = nil

        self.commandDelegate.send(CDVPluginResult.init(status: .ok), callbackId: command.callbackId)
    }

    @objc(checkIfPendingPushExists:)
    func checkIfPendingPushExists(command: CDVInvokedUrlCommand){
        if (OptimoveSDKPlugin.cordovaCommand == nil){
            OptimoveSDKPlugin.cordovaCommand = command
        }

        let sent = maybeSendAndClearPendingPush(command: command)
        if (!sent){
            let pluginResult: CDVPluginResult = CDVPluginResult(status: .ok)
            pluginResult.setKeepCallbackAs(true)
            self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
        }
    }

    private func maybeSendAndClearPendingPush(command: CDVInvokedUrlCommand) -> Bool{
        guard let notification = OptimoveSDKPlugin.pendingPush else {
            return false
        }

        OptimoveSDKPlugin.optimovePluginInstance.sendMessageToJs(type: "pushOpened", data: OptimoveSDKPlugin.getPushNotificationMap(pushNotification: notification))
        OptimoveSDKPlugin.pendingPush = nil

        return true
    }

    private func sendMessageToJs(type: String, data: [String: Any?]?){
        guard let command = OptimoveSDKPlugin.cordovaCommand else {
            return
        }

        var message: [String: Any] = [
            "type": type
        ]

        if (data != nil){
            message["data"] = data
        }

        let pluginResult: CDVPluginResult = CDVPluginResult(status: .ok, messageAs: message)
        pluginResult.setKeepCallbackAs(true)
        self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
    }

    @objc(pushRequestDeviceToken:)
    func pushRequestDeviceToken(command: CDVInvokedUrlCommand) {
        Optimove.shared.pushRequestDeviceToken()

        self.commandDelegate.send(CDVPluginResult(status: .ok), callbackId: command.callbackId)
    }

    @objc(pushUnregister:)
    func pushUnregister(command: CDVInvokedUrlCommand) {
        Optimove.shared.pushUnregister()

        self.commandDelegate.send(CDVPluginResult(status: .ok), callbackId: command.callbackId)
    }

    @objc(inAppUpdateConsent:)
    func inAppUpdateConsent(command: CDVInvokedUrlCommand) {
        let consented: Bool = command.arguments[0] as! Bool
        OptimoveInApp.updateConsent(forUser: consented)

        self.commandDelegate.send(CDVPluginResult(status: .ok), callbackId: command.callbackId)
    }

    @objc(inAppGetInboxItems:)
    func inAppGetInboxItems(command: CDVInvokedUrlCommand) {
        self.commandDelegate.run(inBackground:{
            let inboxItems = OptimoveInApp.getInboxItems()
            var items = [[String : Any?]]()

            let formatter = ISO8601DateFormatter()
            formatter.timeZone = TimeZone(secondsFromGMT: 0)

            for item in inboxItems {
                let dict: [String: Any?] = [
                    "id": item.id,
                    "title": item.title,
                    "subtitle": item.subtitle,
                    "availableFrom": item.availableFrom != nil ? formatter.string(from: item.availableFrom!) : nil,
                    "availableTo": item.availableTo != nil ? formatter.string(from: item.availableTo!) : nil,
                    "dismissedAt": item.dismissedAt != nil ? formatter.string(from: item.dismissedAt!) : nil,
                    "isRead": item.isRead(),
                    "sentAt": formatter.string(from: item.sentAt),
                    "imageUrl": item.getImageUrl()?.absoluteString,
                    "data": item.data
                ]

                items.append(dict)
            }

            let pluginResult = CDVPluginResult(status: .ok, messageAs: items)
            self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
        })
    }

    @objc(inAppPresentInboxMessage:)
    func inAppPresentInboxMessage(command: CDVInvokedUrlCommand) {
        self.commandDelegate.run(inBackground: {
            let messageId = command.arguments[0] as! Int64
            let inboxItems = OptimoveInApp.getInboxItems()

            var presentationResult: InAppMessagePresentationResult = .FAILED
            for msg in inboxItems {
                if (msg.id == messageId){
                    presentationResult = OptimoveInApp.presentInboxMessage(item: msg)
                    break
                }
            }

            var pluginResult = CDVPluginResult.init(status: .ok, messageAs: presentationResult.rawValue)

            self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
        })
    }

    @objc(inAppDeleteMessageFromInbox:)
    func inAppDeleteMessageFromInbox(command: CDVInvokedUrlCommand) {
        self.commandDelegate.run(inBackground: {
            let messageId = command.arguments[0] as! Int64
            let inboxItems = OptimoveInApp.getInboxItems()

            var pluginResult = CDVPluginResult.init(status: .error, messageAs: "Message not found or not available")
            for msg in inboxItems {
                if msg.id != messageId {
                    continue
                }

                let result = OptimoveInApp.deleteMessageFromInbox(item: msg)
                if result {
                    pluginResult = CDVPluginResult.init(status: .ok)
                }
                break
            }

            self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
        })
    }

    @objc(inAppMarkAsRead:)
    func inAppMarkAsRead(command: CDVInvokedUrlCommand) {
        self.commandDelegate.run(inBackground: {
            let messageId = command.arguments[0] as! Int64
            let inboxItems = OptimoveInApp.getInboxItems()

            var pluginResult = CDVPluginResult.init(status: .error, messageAs: "Message not found")
            for msg in inboxItems {
                if msg.id != messageId {
                    continue
                }

                let result = OptimoveInApp.markAsRead(item: msg)
                pluginResult = result ?
                CDVPluginResult.init(status: .ok) :
                CDVPluginResult.init(status: .error, messageAs: "Failed to mark message as read")

                break
            }
            self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
        })
    }

    @objc(inAppMarkAllInboxItemsAsRead:)
    func inAppMarkAllInboxItemsAsRead(command: CDVInvokedUrlCommand) {
        self.commandDelegate.run(inBackground: {
            let result = OptimoveInApp.markAllInboxItemsAsRead()

            if result {
                self.commandDelegate.send(.init(status: .ok), callbackId: command.callbackId)
            }
            else {
                self.commandDelegate.send(.init(status: .error, messageAs: "Failed to mark all messages as read"), callbackId: command.callbackId)
            }
        })
    }

    @objc(inAppGetInboxSummary:)
    func inAppGetInboxSummary(command: CDVInvokedUrlCommand) {
        OptimoveInApp.getInboxSummaryAsync { summary in
            if let summary = summary {
                self.commandDelegate.send(.init(status: .ok, messageAs: ["totalCount" : summary.totalCount, "unreadCount" : summary.unreadCount]), callbackId: command.callbackId)
            }
            else {
                self.commandDelegate.send(.init(status: .error, messageAs: "Could not get inbox summary"), callbackId: command.callbackId)
            }
        }
    }

    private static func getPushNotificationMap(pushNotification: PushNotification) -> [String: Any?] {
        let aps: [AnyHashable:Any] = pushNotification.aps
        var alert: [String: String] = [:]
        if let a = aps["alert"] as? Dictionary<String, String> {
            alert = a
        }

        let title: String? = alert["title"] ?? nil
        let message: String? = alert["body"] ?? nil

        let dict: [String: Any?] = [
            "id": pushNotification.id,
            "title": title,
            "message": message,
            "data": pushNotification.data,
            "url": pushNotification.url?.absoluteString,
            "actionId": pushNotification.actionIdentifier
        ]

        return dict
    }

    private static func getInappButtonPressMap(inAppButtonPress: InAppButtonPress) -> [String: Any?] {
        let dict: [String: Any?] = [
            "deepLinkData": inAppButtonPress.deepLinkData,
            "messageData": inAppButtonPress.messageData,
            "messageId": inAppButtonPress.messageId
        ]

        return dict
    }


    // ========================== DDL ==========================

    @objc(checkIfPendingDDLExists:)
    func checkIfPendingDDLExists(command: CDVInvokedUrlCommand){
        if (OptimoveSDKPlugin.cordovaCommand == nil){
            OptimoveSDKPlugin.cordovaCommand = command
        }

        let sent = maybeSendAndClearPendingDdl(command: command)
        if (!sent){
            let pluginResult: CDVPluginResult = CDVPluginResult(status: .ok)
            pluginResult.setKeepCallbackAs(true)
            self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
        }
    }

    private func maybeSendAndClearPendingDdl(command: CDVInvokedUrlCommand) -> Bool{
        guard let ddlResolution = OptimoveSDKPlugin.pendingDdl else {
            return false
        }

        OptimoveSDKPlugin.optimovePluginInstance.sendMessageToJs(type: "deepLink", data: OptimoveSDKPlugin.getDdlResolutionMap(deepLinkResolution: ddlResolution))
        OptimoveSDKPlugin.pendingDdl = nil

        return true
    }

    private static func getDdlResolutionMap(deepLinkResolution: DeepLinkResolution) -> [String: Any?] {
        var urlString: String
        var resolution: String
        var content: [String: Any?]? = nil
        var linkData: [AnyHashable:Any?]? = nil

        switch deepLinkResolution {
            case .lookupFailed(let dl):
                urlString = dl.absoluteString
                resolution = "LOOKUP_FAILED"
            case .linkNotFound(let dl):
                urlString = dl.absoluteString
                resolution = "LINK_NOT_FOUND"
            case .linkExpired(let dl):
                urlString = dl.absoluteString
                resolution = "LINK_EXPIRED"
            case .linkLimitExceeded(let dl):
                urlString = dl.absoluteString
                resolution = "LINK_LIMIT_EXCEEDED"
            case .linkMatched(let dl):
                urlString = dl.url.absoluteString
                resolution = "LINK_MATCHED"
                content = [
                    "title": dl.content.title,
                    "description": dl.content.description,
                ]
                linkData = dl.data
        }

        return [
            "resolution": resolution,
            "url": urlString,
            "content": content,
            "linkData": linkData,
        ]
    }

    @objc(application:userActivity:restorationHandler:)
    static func application(_ application: UIApplication, userActivity: NSUserActivity, restorationHandler: @escaping ([UIUserActivityRestoring]?) -> Void){
        _ = Optimove.shared.application(application, continue: userActivity, restorationHandler: restorationHandler)
    }

    @available(iOS 13.0, *)
    @objc(scene:session:options:)
    static func scene(_ scene: UIScene, session: UISceneSession, options: UIScene.ConnectionOptions) {
        guard let _ = (scene as? UIWindowScene) else { return }

        // Deep links from cold starts
        if let userActivity = options.userActivities.first {
            Optimove.shared.scene(scene, continue: userActivity)
        }
    }

    @available(iOS 13.0, *)
    @objc(scene:userActivity:)
    static func scene(_ scene: UIScene, userActivity: NSUserActivity) {
        Optimove.shared.scene(scene, continue: userActivity)
    }
}
